File: api/token_crypto.py
"""AES-256-GCM token encryption at rest — port of openreply's
lib/meta/oauth.ts encryptToken/decryptToken (MIT License,
Copyright (c) 2026 Anish Raj, Diwen Huang).

Format: "enc1:" || base64url(nonce[12] || ciphertext || tag).
Key: 32-byte base64url in TOKEN_ENCRYPTION_KEY. Generate:

    python -c "import base64,os;print(base64.urlsafe_b64encode(os.urandom(32)).decode())"
"""
from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_PREFIX = "enc1:"


class TokenCrypto:
    """AES-256-GCM encrypt/decrypt with an "enc1:" prefix for migration safety."""

    def __init__(self, key_b64: str) -> None:
        self._aesgcm = AESGCM(base64.urlsafe_b64decode(key_b64.encode()))

    def encrypt(self, plaintext: str) -> str:
        """Return "enc1:" || base64url(nonce || ciphertext || tag)."""
        nonce = os.urandom(12)
        ct = self._aesgcm.encrypt(nonce, plaintext.encode(), None)
        return _PREFIX + base64.urlsafe_b64encode(nonce + ct).decode()

    def decrypt(self, stored: str) -> str:
        """Strip "enc1:" prefix, then base64url-decode and decrypt."""
        raw = base64.urlsafe_b64decode(stored.removeprefix(_PREFIX).encode())
        return self._aesgcm.decrypt(raw[:12], raw[12:], None).decode()


def get_token_crypto() -> TokenCrypto:
    """Build a TokenCrypto from the TOKEN_ENCRYPTION_KEY env var."""
    key = os.getenv("TOKEN_ENCRYPTION_KEY", "")
    if not key:
        raise RuntimeError("TOKEN_ENCRYPTION_KEY env var is not set")
    return TokenCrypto(key)


def decrypt_or_plaintext(stored: str, crypto: TokenCrypto) -> str:
    """Migration helper: legacy rows hold plaintext tokens.

    Attempts to decrypt; if decryption fails (legacy plaintext or
    corrupted value), returns the original string unchanged.
    """
    try:
        return crypto.decrypt(stored)
    except Exception:
        return stored

File: api-go/internal/platform/crypto/token_crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"fmt"
)

const prefix = "enc1:"

type TokenCrypto struct {
	gcm cipher.AEAD
}

func New(keyB64 string) (*TokenCrypto, error) {
	key, err := base64.URLEncoding.DecodeString(keyB64)
	if err != nil {
		key, err = base64.RawURLEncoding.DecodeString(keyB64)
		if err != nil {
			return nil, fmt.Errorf("decode key: %w", err)
		}
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("new gcm: %w", err)
	}

	return &TokenCrypto{gcm: gcm}, nil
}

func (t *TokenCrypto) Encrypt(plaintext string) (string, error) {
	nonce := make([]byte, 12)
	if _, err := rand.Read(nonce); err != nil {
		return "", fmt.Errorf("generate nonce: %w", err)
	}

	ciphertext := t.gcm.Seal(nil, nonce, []byte(plaintext), nil)
	return prefix + base64.RawURLEncoding.EncodeToString(append(nonce, ciphertext...)), nil
}

func (t *TokenCrypto) Decrypt(stored string) (string, error) {
	raw, err := base64.RawURLEncoding.DecodeString(trimPrefix(stored))
	if err != nil {
		return "", fmt.Errorf("decode ciphertext: %w", err)
	}
	if len(raw) < 13 {
		return "", fmt.Errorf("ciphertext too short")
	}

	plaintext, err := t.gcm.Open(nil, raw[:12], raw[12:], nil)
	if err != nil {
		return "", fmt.Errorf("decrypt token: %w", err)
	}
	return string(plaintext), nil
}

func DecryptOrPlaintext(stored string, crypto *TokenCrypto) string {
	if crypto == nil {
		return stored
	}
	plaintext, err := crypto.Decrypt(stored)
	if err != nil {
		return stored
	}
	return plaintext
}

// DecryptOrPlaintext decrypts enc1: tokens or returns the stored value unchanged
// for legacy plaintext tokens (implements worker.TokenDecryptor / handlers.TokenCrypto).
func (t *TokenCrypto) DecryptOrPlaintext(stored string) string {
	return DecryptOrPlaintext(stored, t)
}

func trimPrefix(stored string) string {
	if len(stored) >= len(prefix) && stored[:len(prefix)] == prefix {
		return stored[len(prefix):]
	}
	return stored
}
